### ISSUE TITLE
NIX_SSHOPTS not passed properly to nix-build in nixos-rebuild.sh 

### ISSUE THREAD

**Issue by wamserma** [NixOS/nixpkgs#80142]
**Describe the bug**

When running 
```
export NIX_SSHOPTS="-i /root/.ssh/id_nix_store"
nixos-rebuild switch --upgrade --option extra-binary-caches ssh://nix-ssh@server
```
no public key authentication happens (as shown in log of sshd on server)
stracing showed me that this seems to be the failing call
```
execve("/tmp/nixos-rebuild.Rvox9Q/nix/bin/nix-build", ["nix-build", "<nixpkgs/nixos>", "--no-out-link", "-A", "system", "--option", "extra-binary-caches", "ssh://nix-ssh@server"], 0x1ca2008 /* 56 vars */) = 0
```

I first suspected this to be a `nix`issue, hence my report at https://github.com/NixOS/nix/issues/3314

**To Reproduce**
please see https://github.com/NixOS/nix/issues/3314

**Expected behavior**
nixos-rebuiöd should fetch all missing binaries from the ssh-served store, then rebuild the system.

**Additional context**

**Metadata**
Client
```
- system: `"x86_64-linux"`
 - host os: `Linux 4.19.101, NixOS, 19.09.2070.b9cb3b2fb2f (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.2`
 - channels(root): `"nixos-19.09.2070.b9cb3b2fb2f, unstable-20.03pre212208.8130f3c1c2b"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Server
```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.102, NixOS, 19.09.2070.b9cb3b2fb2f (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.2`
 - channels(root): `"nixos-19.09.2070.b9cb3b2fb2f, unstable-20.03pre212208.8130f3c1c2b"`
 - channels(markus): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: system
# a list of nixos modules affected by the problem
module:
```

**Comment by danbst** [comment]
reading through https://github.com/NixOS/nix/issues/1994 it seems NIX_SSHOPTS isn't supported for `ssh://` paths.
The best thing you could do would to create custom ssh_config and pass as a parameter, described in https://github.com/NixOS/nix/issues/1994#issuecomment-392195357

**Comment by wamserma** [comment]
> reading through [NixOS/nix#1994](https://github.com/NixOS/nix/issues/1994) it seems NIX_SSHOPTS isn't supported for `ssh://` paths.
> The best thing you could do would to create custom ssh_config and pass as a parameter, described in [NixOS/nix#1994 (comment)](https://github.com/NixOS/nix/issues/1994#issuecomment-392195357)

No, `NIX_SSHOPTS` works fine when `nix-build`/ `nix build` is invoked directly. https://github.com/NixOS/nix/issues/1994 is about passing the port number in the `ssh://`-URL instead of `NIX_SSHOPTS`. This is issue is about `NIX_SSHOPTS` somehow not getting properly passed to `nix-build`/ `nix build` when called through `nixos-rebuild`. I tried some debbuging with local nixpkgs (because `nixos-rebuild` re-launches itself after updating) and tracing (ob bash and strace level) on and so forth, but couldn't pinpoint, less fix the issue on my own.

**Comment by danbst** [comment]
ok. I've debugged the issue and figure out that NIX_SSHOPTS passes fine up to inner `nix-build`. There may be issues when you use remote nixos operations, but for stock `switch` it passes fine.

Can you try `nix-build '<nixpkgs/nixos>' -A system --option extra-binary-caches ssh://nix-ssh@server` with your NIX_SSHOPTS hook?

For me, NIX_SSHOPTS doesn't work at all. What works, is setting ssh host in `/etc/ssh/ssh_config`:
```
Host remote_host
  IdentityFile /root/.ssh/private.key
  User your_user
```

**Comment by wamserma** [comment]
> Can you try
> 
> ```
> nix-build '<nixpkgs/nixos>' -A system \
>   --option extra-binary-caches ssh://nix-ssh@server \
>   --option builders ssh://nix-ssh@server \
>   -j 0
> ```
> 
> with your NIX_SSHOPTS hook?

This works fine, just as expected. In fact my current workaround is a snippet of bash doing 
```bash
export NIX_SSHOPTS="-i /root/.ssh/id_nix_store -o ServerAliveInterval=15 -o ServerAliveCountMax=3000 -o IPQoS=throughput"

nix-channel --update
nix build  --no-link '(with import <nixpkgs/nixos> { }; system)' --option extra-binary-caches ssh://nix-ssh@server
nixos-rebuild switch
```
 
> For me, NIX_SSHOPTS doesn't work at all. What works, is setting ssh host in `/etc/ssh/ssh_config`:

did you mean `/root/.ssh/ssh_config`?

> ```
> Host remote_host
>   IdentityFile /root/.ssh/private.key
>   User your_user
> ```

Well that didn't make a difference. On the server I still see:

```
Feb 18 --:--:-- server sshd[2767]: Connection from 192.168.XXX.XXX port 36508 on 192.168.YYY.YYY port 22
Feb 18 --:--:-- server sshd[2767]: error: PAM: Authentication failure for nix-ssh from 192.168.XXX.XXX
Feb 18 --:--:-- server sshd[2767]: Failed keyboard-interactive/pam for nix-ssh from 192.168.XXX.XXX port 36508 ssh2
Feb 18 --:--:-- server sshd[2767]: Connection closed by authenticating user nix-ssh 192.168.XXX.XXX port 36508 [preauth]
```

but when calling `nix-build` directly I can observe
```

**Comment by danbst** [comment]
okay, I give up. `NIX_SSHOPTS` doesn't work at all for me, I can't pass any SSH arguments even to `nix-build`. `/root/.ssh/config` and `/etc/ssh/ssh_config` both work for me though.

**Comment by wamserma** [comment]
That, in turn, seems strange to me as I have a pretty standard out-of-the box setup.

**Comment by imincik** [comment]
In case it helps to others, I use following command for deployment using no-standard SSH port:

```
NIX_SSHOPTS="-p <SSH-PORT>" nixos-rebuild switch --use-remote-sudo --fast --flake .#<HOST-NAME> --target-host <USER>@<HOST>
```

`<USER>` must exist in `nix.settings.trusted-users = [ <USER> <@GROUP> ];` on target host.

Using NixOS 22.11

**Comment by fusion44** [comment]
I have the same problem. I've followed [this guide](https://nix.dev/tutorials/nixos/distributed-builds-setup.html) to setup the remote builder. `nixos-rebuild switch` doesn't use the builder either automatically nor when I pass `--build-host remotebuild@buildhost --use-remote-sudo`.  Setting `NIX_SSHOPTS` as described above won't help either.

As a workaround, I do a manual build of my config and then do the switch:
`nix build --max-jobs 0 --impure .#nixosConfigurations.myConfig.config.system.build.toplevel`
